### DISCUSSION TITLE
Issues with viewing the pass (on mobile, and on mac preview)

### DISCUSSION THREAD

**Discussion by uhel-dev** [alexandercerutti/passkit-generator#208] [Q&A]
Hi there, first of all time you've spend on doing this 100% worth it, helps a lot of us get thru this pain. Thank you.

I have an odd one, I managed to get thru certs, had issue with the passphrase solved that, now the function creates the pass but is unusable, do you have an idea what it could be 

```javascript
    PKPass.from({
        model: "./Generic.pass",
        certificates: {
            wwdr: fs.readFileSync("certs/wwdr.pem"),
            signerCert: fs.readFileSync("certs/signerCert.pem"),
            signerKey: fs.readFileSync("certs/signerKey.pem"),
            signerKeyPassphrase: "signerKeyPassphrase!",
        }
    }).then(async (newPass) => {

            // Save the pass to a file
            const buffer = await newPass.getAsBuffer();
            fs.writeFileSync('pass.pkpass', buffer);

            res.status(200).send('Pass generated and saved as custom_pass.pkpass');
        }).catch(error => {
            console.error("Error generating pass:", error);
            res.status(500).send('Error generating pass');
        });
```

Even when trying to build a Generic pass with no changes, it doesn't open, when I do the same with the signpass from apple it creates usable pass, I'm sure I'm missing something tiny, but been at it for too long to give up, let me know if you have any ideas.

Best,

**Comment by alexandercerutti** [comment]
Hey @uhel-dev! Thanks for using passkit-generator! Glad this is helpful.

Did you try to debug through Console.app? I created a guide in wiki for that, in order to find the message: https://github.com/alexandercerutti/passkit-generator/wiki/Troubleshooting-(Self-help)#through-console-mac-only

Once you are able to check the message, we can dig down for the reason.

Let me know.

**Reply by uhel-dev** [reply]
Man is a legend, replying so fast @alexandercerutti thank you. I have tried debugging but no luck yet, seen the amazing troubleshooting but not sure if I know what I'm looking for, so you scan logs of iphone not the mac right? thats what im doing now

**Reply by uhel-dev** [reply]
ah I think I found something, now just need to figure out what it is 😓 
```
Verifying structure for pass <private>
Verifying signature for pass <private> using network:1
signature verification failed for signer 0
CMS signature verification failed: Error Domain=MSErrorCMSDomain Code=-67808 "signature verification failed for signer 0" UserInfo={NSLocalizedDescription=signature verification failed for signer 0}
CMS verification error: -25293
Signature validation: *** FAILED ***
Invalid data error reading pass pass.com.apple.devpubs.example/8j23fm3. Manifest signature did not verify successfully
Could not generate thumbnail for invalid pass. Error: Error Domain=PKPassKitErrorDomain Code=1 UserInfo={NSLocalizedDescription=<private>, NSUnderlyingError=0xaea3707b0 {Error Domain=PKPassKitErrorDomain Code=1 UserInfo={NSLocalizedDescription=<private>}}}
com.apple.quicklook.ThumbnailsAgent	Error while generating thumbnail with third party extension: Error Domain=PKPassKitErrorDomain Code=1 UserInfo={NSLocalizedDescription=<private>, NSUnderlyingError=0xada3b7bd0 {Error Domain=PKPassKitErrorDomain Code=1 UserInfo={NSLocalizedDescription=<private>}}}. #Thumbnail
```

**Reply by alexandercerutti** [reply]
That's what we are looking for. Great job!

You seem to be lacking of a `thumbnail.png` or `thumbnail@2x.png` file, from what I'm understanding. Is it available in your model?

**Reply by uhel-dev** [reply]
![Screenshot 2024-07-06 at 22 49 53](https://github.com/alexandercerutti/passkit-generator/assets/60184122/738efb70-eb40-4df6-a95e-d501b0c081f8)
I do have it in the Generic.pass folder, i think it was wrong identifier, but I solved that and it still doesn't open let me check the logs again

**Reply by uhel-dev** [reply]
```
default	22:51:31.230050+0100	PassKitThumbnailExtension	[0xaea3a4280] Re-initialization successful; calling out to event handler with XPC_ERROR_CONNECTION_INTERRUPTED
default	22:51:31.230220+0100	PassKitThumbnailExtension	[0xaea210000] invalidated after getting a no-senders notification - client is gone
````

so far what I'm seeing

**Reply by alexandercerutti** [reply]
Please note that, depending on the device you are opening your pass on, a different file could be required. For example, newer iPhones require icon 2x or maybe 3x but Macs require 2x.

**Reply by alexandercerutti** [reply]
No, that doesn't help. Keep looking

**Reply by uhel-dev** [reply]
`Manifest signature did not verify successfully`

**Reply by uhel-dev** [reply]
```CMS signature verification failed: Error Domain=MSErrorCMSDomain Code=-67808 "signature verification failed for signer 0" UserInfo={NSLocalizedDescription=signature verification failed for signer 0}
CMS verification error: -25293
Signature validation: *** FAILED ***
Invalid data error reading pass pass.com.something.secret/SERIAL_NUMBER. Manifest signature did not verify successfully
Could not generate thumbnail for invalid pass. Error: Error Domain=PKPassKitErrorDomain Code=1 UserInfo={NSLocalizedDescription=<private>, NSUnderlyingError=0xaea3a9050 {Error Domain=PKPassKitErrorDomain Code=1 UserInfo={NSLocalizedDescription=<private>}}}
com.apple.quicklook.ThumbnailsAgent	Error while generating thumbnail with third party extension: Error Domain=PKPassKitErrorDomain Code=1 UserInfo={NSLocalizedDescription=<private>, NSUnderlyingError=0xada3b6ee0 {Error Domain=PKPassKitErrorDomain Code=1 UserInfo={NSLocalizedDescription=<private>}}}. #Thumbnail
Generating thumbnail for file URL: <private>```

**Reply by uhel-dev** [reply]
I think I'm gonna dig into it more, I just gave up before after not seeing any logs, but thanks to you I found them. Check whenever free and if you see me still stuck we can look into it maybe. I will make sure to update here for visibility

**Reply by alexandercerutti** [reply]
That's nasty... what happens if you unzip the pkpass file? Are you able to open individual files?

**Reply by uhel-dev** [reply]
serial number, pass identifier, team identifier all this matches and work with the signpass, not sure how I would update the manifest?

**Reply by uhel-dev** [reply]
It's a single pkpass, you mean to unzip it with terminal right?

**Reply by alexandercerutti** [reply]
I would guess that it is not a problem with the manifest itself, but something depending on the OS you are using. Sometimes these things happen.

You can rename .pkpass to .zip and unzip it. You can use the terminal or the system UI, as you prefer.

However, I do never exclude the possibility of a bug somewhere in the toolchain. Could be on node-forge, could be in Node, could be in passkit-generator. Sometimes things are not easy to solve

**Reply by alexandercerutti** [reply]
Another thing I could suggest you is to try to not save the pass on your OS through `fs.writeFile` but, instead, serve directly from the browser as part of the HTTP Response. If there's a problem with writing a file through Node, this could help avoiding it.

Like I do here: https://github.com/alexandercerutti/passkit-generator/blob/bab4117cb029b00a50077262187fcc2ad9412357/examples/self-hosted/src/setBarcodes.ts#L70-L77

**Reply by uhel-dev** [reply]
ah you right, I think is the hopelessness now :D need a break too much coffee, I did change to zip unzipped it and see the pass files
![Screenshot 2024-07-06 at 23 03 19](https://github.com/alexandercerutti/passkit-generator/assets/60184122/625c703a-bb51-40a8-ad8e-269eef787e0a)
interesting same file works with the signpass so I can exclude the issue with model, I think you are right, something in between is wrong prolly on my local, gonna get to it tho, keystore I think is set correctly with all the certs. It won't let me co-exist until i get this working :D once again thanks for all the replies man.

**Reply by uhel-dev** [reply]
ah great hint with the `fs.writeFile` it will help me exclude another area of error

**Reply by alexandercerutti** [reply]
Let me know if the situation changes by removing `writeFile`. So, apparently the OS seems to be able to open the files there, but I see files have no "modified date" (which, I guess, makes the OS to fallback to 30 Nov 1979). I'm not sure this could be an issue...

**Reply by uhel-dev** [reply]
serving from browser didn't help, the pass still doesn't open. I'm checking logs now to see if they changed, tl;dr date of the added file is today 23:30 so it could be I didn't provide date indicator when writing the file maybe 🤷

**Reply by alexandercerutti** [reply]
If you want to share the pkpass file, I can perhaps try to give it a look.

**Reply by uhel-dev** [reply]
Your kindness is limitless, let me try to figure it out today (and share outcomes) if that doesn't work out, I will share it and hope for some help. Have a lovely day ahead man.

**Reply by alexandercerutti** [reply]
No. I'm going to have a lovely night instead ahah
You made me realize I didn't have the setting to show my timezone on my profile ahah.

Sure let me know

**Reply by uhel-dev** [reply]
🎉 at least both of us could learn something, lovely night for both of us then, lets see how it gets

**Reply by alexandercerutti** [reply]
I guess you solved in the end, isn't it? Do you perhaps remember what was the issue?

**Reply by uhel-dev** [reply]
Hi there, thats still on my list, been buried with other stuff, we can close it and I can keep this as a reference, because I will be back at it, and defo share the problem. 💯
